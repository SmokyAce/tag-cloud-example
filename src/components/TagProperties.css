.properties {
    padding: 0;
    margin-top: 50px;
    max-width: 400px;
    align-self: center;
    text-align: left;
}

.properties .nav {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 30px;
}

.properties .nav a {
    font-size: 1.3em;
    color: #3e77aa;
    text-decoration: none;
}

.properties .nav a:hover {
    text-decoration: underline;
}

.properties .subheader {
    margin: 5px;
    font-size: 1.1em;
}


.list .subheader {
    list-style-type: none;
    font-size: 1.1em;
    left: -40px;
    position: relative;
}

.list-item {
    padding: 5px 0;
}
