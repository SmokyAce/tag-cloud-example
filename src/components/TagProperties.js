import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect, Link } from 'react-router-dom';
import Spinner from './Spinner';
import { getTagInfoById, getNeighbors } from '../utils/api';
import './TagProperties.css';

class TagProperties extends Component {
  state = {
    tagInfo: {},
    neighbors: {}
  };

  componentDidMount() {
    this.getData();
  }

  componentDidUpdate(prevProps) {
    if (this.props.tagId !== prevProps.tagId) {
      this.getData();
    }
  }

  render() {
    const { tagInfo, neighbors } = this.state;

    if (!tagInfo) {
      return <Redirect to='/' />;
    }

    if (!tagInfo.id) {
      return <Spinner />;
    }

    return (
      <div className='properties'>
        <div className='nav'>
          <Link to={`/home/${neighbors.prev}`}>&#x1F880;</Link>
          <Link className='text-center' to='/home'>
            Home
          </Link>
          <Link to={`/home/${neighbors.next}`}>&#x1F882;</Link>
        </div>
        <h3 className='subheader'>
          {this.upperCaseFirstLetter(tagInfo.label)}
        </h3>
        <ul className='list'>
          <li className='list-item subheader'>
            Total Mentions: {this.getTotalMentions()}
          </li>
          <li className='list-item green'>
            Positive Mentions: {this.getMention('positive')}
          </li>
          <li className='list-item blue'>
            Neutral Mentions: {this.getMention('neutral')}
          </li>
          <li className='list-item red'>
            Negative Mentions: {this.getMention('negative')}
          </li>
          <li className='list-item subheader'>List of page types</li>
          {tagInfo.pageType &&
            Object.keys(tagInfo.pageType).map((type, i) => {
              return (
                <li key={i} className='list-item'>
                  {type}: {tagInfo.pageType[type]}
                </li>
              );
            })}
        </ul>
      </div>
    );
  }

  getData = () => {
    const { tagId } = this.props;
    if (tagId) {
      Promise.all([getTagInfoById(tagId), getNeighbors(tagId)])
        .then(result => {
          if (!result) throw new Error();

          const [tagInfo, neighbors] = result;
          this.setState({
            tagInfo,
            neighbors
          });
        })
        // eslint-disable-next-line
        .catch(console.log.bind(console));
    }
  };

  getTotalMentions = () => {
    let { tagInfo, total = 0 } = this.state;
    if (tagInfo.sentiment) {
      Object.keys(tagInfo.sentiment).forEach(type => {
        total += tagInfo.sentiment[type];
      });
    }
    return total;
  };

  getMention = type => {
    const { tagInfo } = this.state;
    if (tagInfo.sentiment && tagInfo.sentiment[type]) {
      return tagInfo.sentiment[type];
    }
    return 0;
  };

  upperCaseFirstLetter = string => {
    return string.charAt(0).toUpperCase() + string.slice(1);
  };
}

TagProperties.propTypes = {
  tagId: PropTypes.string.isRequired
};

export default TagProperties;
