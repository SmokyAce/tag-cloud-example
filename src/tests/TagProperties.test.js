/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import TagProperties from '../components/TagProperties';

const mockProps = {
  tagId: '1751295897__Odessa'
};

describe('<TagProperties />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TagProperties {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  const testRenderer = TestRenderer.create(<TagProperties {...mockProps} />);

  it('renders correctly', () => {
    const snapshot = testRenderer.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
