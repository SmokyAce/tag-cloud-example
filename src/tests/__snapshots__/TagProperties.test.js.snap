// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagProperties /> renders correctly 1`] = `
<div
  id="loading"
/>
`;
